### PR TITLE
Don't use Git Submodules in base Trilinos repo for GenConfig repos

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This PR changes the handling of the GenConfig repos so as not to mess up the base Trilinos repo and avoid confusion and mistakes by Trilinos developers.  The instructions for using `packages/framework/get_dependencies.sh` are identical, but now it will not show modified state in the local Trilinos repo and avoid mistakes by regular Trilinos developers.

Note that you can easily see the state of these extra GenConfig repos by running:

```
$ cd Trilinos/
$ source cmake/tribits/python_utils/gitdist-setup.sh
$ gitdist-status
```

This has the exact same behavior of the old `packages/framework/get_dependencies.sh` implementation based on Git Submodules in the base Trilinos repo in that it pulls the tip of the 'master' branches for the subrepos `GenConfig`, `son-ini-files`, and `srn-ini-files` and updates the Git Submodules for the GenConfig repo.  However, again, this new implementation does not mess up the state of the base Trilinos Git repo or even make these subrepos visible to regular Trilinos developers unless they want to see them (i.e. with `gitdist`).

By moving away from the old Git Submodules implementation that created a dirty state in the base Trilinos Git repo, we can avoid the following mistakes that some Trilinos developers are sure to make (especially if they are not very familiar with Git Submodules):

* They will accidentally commit the Git submodules versions (and perhaps also changes in the submodules themselves due to things like [TRILINOSHD-102](https://sems-atlassian-son.sandia.gov/jira/servicedesk/customer/portal/7/TRILINOSHD-102)).
* They will forgot to push the updated subrepos (which invalidates the base Git submodule commit).
* Assuming they do push to the repos correctly (highly unlikely), there will be conflicts in the Git submodules commits when different PRs try to merge to ‘develop’ when two non-overlapping PRs are modified in this way.

NOTE: The following MRs should be merged as well so you don't get modified GenConfig repos locally when running `packages/framework/get_dependencies.sh`:

* https://gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/LoadEnv/-/merge_requests/90
* https://gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/GenConfig/-/merge_requests/88

and those MRs must be merged in that order due to the usage of Git Submodules in the GenConfig repo.

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
NOTE: My moving away for the old Git Submodules implementation that created a dirty state in the base Trilinos Git repo, we can avoid the following mistakes that some Trilinos developers are sure to make (especially if they are not very familiar with Git Submodules):

* They will accidentally commit the Git submodules versions (and perhaps also changes in the submodules themselves due to things like [TRILINOSHD-102](https://sems-atlassian-son.sandia.gov/jira/servicedesk/customer/portal/7/TRILINOSHD-102)).
* They will forgot to push the updated subrepos (which invalidates the base Git submodule commit).
* Assuming they do push to the repos correctly (highly unlikely), there will be conflicts in the Git submodules commits when different PRs try to merge to ‘develop’ when two non-overlapping PRs are modified in this way.

File: packages/framework/get_dependencies.sh [hunk 1]
```diff
@@ -6,34 +6,69 @@ pushd $PWD
...
-elif [[ ! -z "$(ls ./son-ini-files)" && "$ini_file_option" == "--son" ]]; then
-    cd GenConfig/deps/LoadEnv/ini_files
+elif [[ -d ${script_dir}/son-ini-files ]] && [[ "$ini_file_option" == "--son" ]]; then
+    echo "Link files from son-ini-files"
     ln -sf ${script_dir}/son-ini-files/trilinos/framework/environment-specs.ini
     ln -sf ${script_dir}/son-ini-files/trilinos/framework/supported-systems.ini
     ln -sf ${script_dir}/ini-files/supported-envs.ini
 else
-    cd GenConfig/deps/LoadEnv/ini_files
+    echo "Link files from init-files"
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:63]:
```suggestion
    echo "Link files from ini-files"
```

**Review by e10harvey — CHANGES_REQUESTED** [review]
Thank you, Ross! Only outstanding issue is adding a GenConfig.sha file (or similar approach via a get_dependencies.sh variable) to ensure we pull in a vetted version of GenConfig and friends.

File: packages/framework/get_dependencies.sh [hunk 2]
```diff
@@ -6,34 +6,69 @@ pushd $PWD
...
     [ -e ${script_dir}/ini-files/environment-specs.ini ] && ln -sf ${script_dir}/ini-files/environment-specs.ini
     [ -e ${script_dir}/ini-files/supported-systems.ini ] && ln -sf ${script_dir}/ini-files/supported-systems.ini
     ln -sf ${script_dir}/ini-files/supported-envs.ini
 fi
+cd - > /dev/null
+cd GenConfig/ini_files
 
-cd ../../../ini_files
 ln -sf ${script_dir}/ini-files/config-specs.ini
 ln -sf ${script_dir}/ini-files/supported-config-flags.ini
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:73]:
```suggestion
# Print summary of ini file settings
cd ${script_dir}
echo "You selected the following LoadEnv ini files:"
ls -lat GenConfig/deps/LoadEnv/ini_files
echo ""
echo "You selected the following GenConfig ini files:"
ls -lat GenConfig/ini_files
echo "If these symlinks do not point to the desired ini files, please re-run:
echo "    $0 [--son, --srn]"
```

**Comment by bartlettroscoe** [comment]
> Thank you, Ross! Only outstanding issue is adding a GenConfig.sha file (or similar approach via a get_dependencies.sh variable) to ensure we pull in a vetted version of GenConfig and friends.

Note that the old implementation was just pulling the tip of the 'master' branch in the three repos.  The updated implementation in this PR does the same thing (without messing with the state of the base Trilinos Git repo).  The fact that things got committed to the 'master' branch of those tree repos should mean that the versions are "vetted".  (That is what the 'master' branch means.)

@e10harvey, If you don't agree, we should discuss what you are trying to do here and come up with a clean way to do it.  (I have been working with and thinking about multi-repo Git projects and workflows for 15+ years.)

File: packages/framework/get_dependencies.sh [hunk 3]
```diff
@@ -6,34 +6,77 @@ pushd $PWD
 
 cd ${script_dir}
 
-# Update GenConfig submodules
-# NOTE: Add --remote below to pull in the latest tip of the GenConfig tracked branch
-git submodule update --init --recursive ${script_dir}/GenConfig
+function tril_genconfig_clone_or_update_repo() {
+  git_url=$1
+  sub_dir=$2
+  has_submodules=$3
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:13]:
```suggestion
  head_sha=$4
```

File: packages/framework/get_dependencies.sh [hunk 4]
```diff
@@ -6,34 +6,77 @@ pushd $PWD
...
 
-# Update ini file submodules
-git submodule update --init --remote ${script_dir}/srn-ini-files || true
-git submodule update --init --remote ${script_dir}/son-ini-files || true
+  if [[ -e ${sub_dir} ]] ; then
+    cd ${sub_dir}
+    git pull
+    cd - > /dev/null
+  else
+    git clone ${git_url} ${sub_dir}
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:19]:
```suggestion
    git clone ${git_url} ${sub_dir}
    cd ${sub_dir}
```

File: packages/framework/get_dependencies.sh [hunk 5]
```diff
@@ -6,34 +6,77 @@ pushd $PWD
...
+  fi
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:21]:
```suggestion
  if [[ ! -z ${head_sha} ]]; then
    git checkout -f ${head_sha}
  fi
```

File: packages/framework/get_dependencies.sh [hunk 6]
```diff
@@ -6,34 +6,77 @@ pushd $PWD
...
+    echo "ERROR: argument '${has_submodules}' not allowed!  Only 'has-submodules' or ''!"
+    exit 1
+  fi
+}
+
+# Clone or update the repos
+
+tril_genconfig_clone_or_update_repo \
+  git@gitlab-ex.sandia.gov:trilinos-devops-consolidation/code/GenConfig.git \
+  GenConfig  has-submodules
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:36]:
```suggestion
  GenConfig  has-submodules b47cbdb1145a9d13f5bc8d4e7e691bab1dcdb60d
```

**Review by e10harvey — CHANGES_REQUESTED** [review]

File: packages/framework/get_dependencies.sh [hunk 7]
```diff
@@ -6,34 +6,77 @@ pushd $PWD
...
+    cd - > /dev/null
+  else
+    git clone ${git_url} ${sub_dir}
+  fi
 
-# Point to srn ini files if they exist
-if [[ ! -z "$(ls ./srn-ini-files)" && "$ini_file_option" == "--srn" ]]; then
-    cd GenConfig/deps/LoadEnv/ini_files
+  if [[ "${has_submodules}" == "has-submodules" ]] ; then
+    cd ${sub_dir}
```
Comment (e10harvey) [inline packages/framework/get_dependencies.sh:23]:
```suggestion
```

**Comment by bartlettroscoe** [comment]
Hello @e10harvey, all of of these recent suggestions should really be a single commit (otherwise, these are broken separate commits).  Can you post a staged PR against this PR with those changes?  Do you remember how to do that?

**Comment by e10harvey** [comment]
> Hello @e10harvey, all of of these recent suggestions should really be a single commit (otherwise, these are broken separate commits). Can you post a staged PR against this PR with those changes? Do you remember how to do that?

I pushed to your remote by mistake. Please see [06b51e0](https://github.com/trilinos/Trilinos/pull/10620/commits/06b51e0082d569d5390ff3dd4dd018ae3fe64e6b). I will replay a PR pipeline that uses this new script and post the results.

**Comment by e10harvey** [comment]
> > Hello @e10harvey, all of of these recent suggestions should really be a single commit (otherwise, these are broken separate commits). Can you post a staged PR against this PR with those changes? Do you remember how to do that?
> 
> I pushed to your remote by mistake. Please see [06b51e0](https://github.com/trilinos/Trilinos/pull/10620/commits/06b51e0082d569d5390ff3dd4dd018ae3fe64e6b). I will replay a PR pipeline that uses this new script and post the results.

[Results from replay using source branch.](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic141&field2=buildname&compare2=61&value2=PR-10620-test-rhel7_sems-gnu-7.2.0-anaconda3-serial_debug_shared_no-kokkos-arch_no-asan_no-complex_no-fpic_no-mpi_no-pt_no-rdc_no-uvm_deprecated-on_pr-framework-84&field3=buildstamp&compare3=61&value3=20220617-1839-Experimental)

**Review by e10harvey — APPROVED** [review]

File: packages/framework/get_dependencies.sh [hunk 8]
```diff
@@ -7,33 +7,39 @@ pushd $PWD
...
 function tril_genconfig_clone_or_update_repo() {
+  pushd $PWD
   git_url=$1
   sub_dir=$2
   has_submodules=$3
+  head_sha=$4
 
   if [[ -e ${sub_dir} ]] ; then
     cd ${sub_dir}
-    git pull
```
Comment (bartlettroscoe) [inline packages/framework/get_dependencies.sh]: You can't remove the `git pull` for the repos where you did not pass in `head_sha`.  Just doing a `git fetch` does not put the repo in the correct state.  There is a missing `git pull`.  I will add the commit that does this.

**Comment by bartlettroscoe** [comment]
@e10harvey, please review the staggered PR https://github.com/bartlettroscoe/Trilinos/pull/7 which contains what should be final updates to this PR branch.  With that version of the get_dependencies.sh script, I see no modified files in the base Trilinos repo or subrepos.

**Comment by bartlettroscoe** [comment]
Shoot, this still had `AT: AUTOMERGE` set which merged this PR when PR testing finally passed.  I should have removed  `AT: AUTOMERGE` when I created the stacked PR https://github.com/bartlettroscoe/Trilinos/pull/7.  That was my fault.  The new follow-on PR #10643.

**Pull request merged**